File: 프로그래머스/lv4/43236. 징검다리/징검다리.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int solution(int distance, vector<int> rocks, int n) {
    int answer = 0;
    sort(rocks.begin(), rocks.end());
    
    int left = 1, right = distance;
    while(left <= right) {
        int middle = (left + right) / 2;
        
        int cnt = 0, prevIdx = 0;
        for(int rock: rocks) {
            if(rock - prevIdx >= middle) {
                prevIdx = rock;
            }
            else{
                cnt++;
            }
        }
        
        if(distance - prevIdx < middle) cnt++;
        
        if(cnt <= n) {
            answer = middle;
            left = middle + 1;
        }
        else{
            right = middle - 1;
        }       
    }
    
    return answer;
}

File: 프로그래머스/4/43236. 징검다리/징검다리.py
def solution(distance, rocks, n):
    answer = 0
    rocks.sort()
    
    start = 1
    end = distance
    while start <= end:
        middle = (start + end) // 2
        
        cnt = 0
        prev_rock = 0
        for rock in rocks:
            if rock - prev_rock >= middle:
                prev_rock = rock
            else:
                # middle이 거리의 최솟값이 될 수 있게, 현재 'rock'은 제거해줘야함
                cnt += 1
                
        if distance - prev_rock < middle:
            cnt += 1
        
        if cnt <= n:
            # 최대 n개의 바위를 제거하면서 최솟값 middle 달성 가능
            answer = middle
            # 최솟값 늘려보기
            start = middle + 1
        else:
            # 달성 불가능, n개보다 더 제거해줘야함
            end = middle - 1
    
    return answer